instrumentation/opentelemetry-instrumentation-fastapi/tests/test_fastapi_instrumentation.py

```diff
@@ -485,6 +580,58 @@ def tearDown(self):
         self._instrumentor.uninstrument()
         super().tearDown()
 
+    def test_sub_app_fastapi_call(self):
```

Comment:
If this is the very same code of the TestAutoInstrumentation we should probably rethink the inheritance of these tests cases

Reply:
something like this 

```
diff --git a/instrumentation/opentelemetry-instrumentation-fastapi/tests/test_fastapi_instrumentation.py b/instrumentation/opentelemetry-instrumentation-fastapi/tests/test_fastapi_instrumentation.py
index 4269dfa2..9c396c66 100644
--- a/instrumentation/opentelemetry-instrumentation-fastapi/tests/test_fastapi_instrumentation.py
+++ b/instrumentation/opentelemetry-instrumentation-fastapi/tests/test_fastapi_instrumentation.py
@@ -61,7 +61,7 @@ _recommended_attrs = {
 }
 
 
-class TestFastAPIManualInstrumentation(TestBase):
+class TestBaseFastAPI:
     def _create_app(self):
         app = self._create_fastapi_app()
         self._instrumentor.instrument_app(
@@ -109,6 +109,30 @@ class TestFastAPIManualInstrumentation(TestBase):
             self._instrumentor.uninstrument()
             self._instrumentor.uninstrument_app(self._app)
 
+    @staticmethod
+    def _create_fastapi_app():
+        app = fastapi.FastAPI()
+
+        @app.get("/foobar")
+        async def _():
+            return {"message": "hello world"}
+
+        @app.get("/user/{username}")
+        async def _(username: str):
+            return {"message": username}
+
+        @app.get("/exclude/{param}")
+        async def _(param: str):
+            return {"message": param}
+
+        @app.get("/healthzz")
+        async def _():
+            return {"message": "ok"}
+
+        return app
+
+
+class TestFastAPIManualInstrumentation(TestBaseFastAPI, TestBase):
     def test_instrument_app_with_instrument(self):
         if not isinstance(self, TestAutoInstrumentation):
             self._instrumentor.instrument()
@@ -310,30 +334,8 @@ class TestFastAPIManualInstrumentation(TestBase):
                 if isinstance(point, NumberDataPoint):
                     self.assertEqual(point.value, 0)
 
-    @staticmethod
-    def _create_fastapi_app():
-        app = fastapi.FastAPI()
-
-        @app.get("/foobar")
-        async def _():
-            return {"message": "hello world"}
-
-        @app.get("/user/{username}")
-        async def _(username: str):
-            return {"message": username}
-
-        @app.get("/exclude/{param}")
-        async def _(param: str):
-            return {"message": param}
-
-        @app.get("/healthzz")
-        async def _():
-            return {"message": "ok"}
-
-        return app
-
 
-class TestFastAPIManualInstrumentationHooks(TestFastAPIManualInstrumentation):
+class TestFastAPIManualInstrumentationHooks(TestBaseFastAPI, TestBase):
     _server_request_hook = None
     _client_request_hook = None
     _client_response_hook = None
@@ -383,7 +385,7 @@ class TestFastAPIManualInstrumentationHooks(TestFastAPIManualInstrumentation):
             )
 
 
-class TestAutoInstrumentation(TestFastAPIManualInstrumentation):
+class TestAutoInstrumentation(TestBaseFastAPI, TestBase):
     """Test the auto-instrumented variant
 
     Extending the manual instrumentation as most test cases apply
@@ -445,7 +447,7 @@ class TestAutoInstrumentation(TestFastAPIManualInstrumentation):
         super().tearDown()
 
 
-class TestAutoInstrumentationHooks(TestFastAPIManualInstrumentationHooks):
+class TestAutoInstrumentationHooks(TestBaseFastAPI, TestBase):
     """
     Test the auto-instrumented variant for request and response hooks
 
```

Reply:
The thing is the cases are not really the same. Because for auto instrumentation you get more spans then for explicit instrumentation (with regards to sub_app usage).

That is the reason why I decided to not split it up but separating only the methods differing.

But I know it seems a bit like repeated code.

Reply:
@xrmx I have completely restructured the unit tests for fastapi instrumentation

1) I moved wrapped and custom header tests to separate py files
2) I restructured the other tests very much like you mentioned